Add tests for increased coverage.

Lussebullen opened the pull request:
**- What I did**

Increased test coverage of `cli/compose/convert/volume.go` to 100%.

Fixed typo in `cli/compose/template/template_test.go` to reach the intended branch test in `extractVariables` func
It checks for `Strings.Contains` func checks for `:?` and not `?:`

Added tests of Descope and Name methods for Namespace in `cli/compose/convert/compose_test.go`.

Added test of convertEnvironment in case of nil value in `cli/compose/convert/service_test.go`.

Added tests for `capitalizeFirst` func in `cli/command/utils_test.go`.

Added tests of String and Type methods for NetworkOpt in `opts/network_test.go`.

Added test for GetAllOrEmpty method and for ParseCPUs in `opts/opts_test.go`.

**- How I did it**

We started by checking the [Code coverage](https://app.codecov.io/gh/docker) to see where testing was lacking and added test as we saw fit in order to increase the test coverage.

**- How to verify it**

Check difference as reported by codecov bot when triggered.

Verify it by running first running `make dev` and `make test` to check that all test passes and afterwards running the following commands to check that the test coverage has increaed:

```bash
$ go test -coverprofile=coverage.out ./...
$ go tool cover -html=coverage.out -o coverage.html
```

or 

`make test-coverage`

**- Description for the changelog**

Add various test cases for increased test coverage.

**- A picture of a cute animal (not mandatory but encouraged)**
![image](https://github.com/docker/cli/assets/64321057/f553b1f2-f0ee-4bb3-9340-f30bff3c7d90)

Lussebullen commented:
The failed tests here do not appear when running ```make test-coverage``` locally, so I cannot reproduce it at this time.
```
 55.07 === Failed
55.07 === FAIL: cli-plugins/socket TestConnectAndWait/connect_goroutine_exits_after_EOF (0.00s)
55.07     socket_test.go:111: assertion failed: 9 (int) != 10 (numGoroutines + 1 int)
55.07 
55.07 === FAIL: cli-plugins/socket TestConnectAndWait (0.00s)
55.07 
55.07 DONE 1719 tests, 1 skipped, 2 failures in 54.735s
------
Dockerfile:77
--------------------
  76 |     ENV GO111MODULE=auto
  77 | >>> RUN --mount=type=bind,target=.,rw \
  78 | >>>     --mount=type=cache,target=/root/.cache \
  79 | >>>     --mount=type=cache,target=/go/pkg/mod \
  80 | >>>     gotestsum -- -coverprofile=/tmp/coverage.txt $(go list ./... | grep -vE '/vendor/|/e2e/')
  81 |     
--------------------
ERROR: failed to solve: process "/bin/sh -c gotestsum -- -coverprofile=/tmp/coverage.txt $(go list ./... | grep -vE '/vendor/|/e2e/')" did not complete successfully: exit code: 1
Error: buildx bake failed with: ERROR: failed to solve: process "/bin/sh -c gotestsum -- -coverprofile=/tmp/coverage.txt $(go list ./... | grep -vE '/vendor/|/e2e/')" did not complete successfully: exit code: 1
```

I'm not entirely sure what to make of the error, but I see that it happens occasionally, e.g. here as well:
https://github.com/docker/cli/pull/4759. Do you know what to make of this?

Benehiko commented:
> The failed tests here do not appear when running `make test-coverage` locally, so I cannot reproduce it at this time.
> 
> ```
>  55.07 === Failed
> 55.07 === FAIL: cli-plugins/socket TestConnectAndWait/connect_goroutine_exits_after_EOF (0.00s)
> 55.07     socket_test.go:111: assertion failed: 9 (int) != 10 (numGoroutines + 1 int)
> 55.07 
> 55.07 === FAIL: cli-plugins/socket TestConnectAndWait (0.00s)
> 55.07 
> 55.07 DONE 1719 tests, 1 skipped, 2 failures in 54.735s
> ------
> Dockerfile:77
> --------------------
>   76 |     ENV GO111MODULE=auto
>   77 | >>> RUN --mount=type=bind,target=.,rw \
>   78 | >>>     --mount=type=cache,target=/root/.cache \
>   79 | >>>     --mount=type=cache,target=/go/pkg/mod \
>   80 | >>>     gotestsum -- -coverprofile=/tmp/coverage.txt $(go list ./... | grep -vE '/vendor/|/e2e/')
>   81 |     
> --------------------
> ERROR: failed to solve: process "/bin/sh -c gotestsum -- -coverprofile=/tmp/coverage.txt $(go list ./... | grep -vE '/vendor/|/e2e/')" did not complete successfully: exit code: 1
> Error: buildx bake failed with: ERROR: failed to solve: process "/bin/sh -c gotestsum -- -coverprofile=/tmp/coverage.txt $(go list ./... | grep -vE '/vendor/|/e2e/')" did not complete successfully: exit code: 1
> ```
> 
> I'm not entirely sure what to make of the error, but I see that it happens occasionally, e.g. here as well: #4759. @thaJeztah do you know what to make of this?

@Lussebullen This is a flaky test unrelated to your changes and can be safely ignored :)

Benehiko commented on cli/command/utils_test.go:
```diff
@@ -68,3 +68,11 @@ func TestValidateOutputPath(t *testing.T) {
 	}
 }
+
+func TestCapitalizeFirstReturnsEmptyStringIfGivenEmptyString(t *testing.T) {
+	assert.Equal(t, capitalizeFirst(""), "")
+}
+
+func TestCapitalizeFirstReturnsStringWithUppercaseFirstLetterIfItWasLowercase(t *testing.T) {
+	assert.Equal(t, capitalizeFirst("blahonga"), "Blahonga")
+}
```
Maybe instead we could use a table test here?

```go

func TestX(t *testing.T) {
  for _, tc := range []struct{
    name string
    val string
    expected string
  } {
    {
      name: "empty string"
      val: "",
      expected: "",
    },
  } {
    t.Run("case="+tc.name, func (t *testing.T) {
      assert.Equal(t, capitalizeFirst(tc.val), tc.expected)
   })
  }
}

Benehiko commented:
Could you resolve the conflicts and the review suggestions? Other than that it looks good! :)

Lussebullen commented on cli/compose/template/template_test.go:
```diff
@@ -220,7 +220,7 @@ func TestExtractVariables(t *testing.T) {
 			name: "required-variable",
 			dict: map[string]any{
-				"foo": "${bar?:foo}",
+				"foo": "${bar:?foo}",
```
Looks like a typo imo, I think the standard parameter expansion syntax uses :?, which is also what is used in all the other assertions in this file. Even if it is not a typo, seems like it would still be best to keep it consistent.

Benehiko replied:
i'm not so sure, the outcome seems to still be the same. I might be wrong here, but from a quick glance it seems any of these `?, :?, ?:` would match here: https://github.com/docker/cli/blob/master/cli/compose/template/template.go#L166-L169 

I would just revert this change

Lussebullen replied:
I reverted the change :)

Lussebullen commented:
I force pushed an amend with my signature to fix the DCO issue, but I'm not sure I can trigger the tests to run again.
I fixed the table test as requested and made a commend on the ?: vs :? section :)

Benehiko commented:
Thank you @Lussebullen, there is still a merge conflict with `cli/command/utils_test.go`. Would you be able to update your branch and fix the merge conflicts?

Lussebullen commented:
@Benehiko reverted the ?: / :? part as you requested, I also updated the branch, but in doing so it seems the test capitalizeFirst utility no longer builds, for some reason it no longer exports despite doing so before. So I removed it entirely.

Benehiko reviewed (approved):
Thanks @Lussebullen, LGTM :)

Lussebullen commented:
My pleasure 🫡
Have a good one ✌️

vvoland reviewed (approved):
LGTM; most of these could be refactored into a table-based tests (thinking of the `convertVolumeToMount` tests).

But this is still an improvement and could be a follow up 😄 

Thanks!

Benehiko commented:
Hey @Lussebullen sorry to ask more of you :see_no_evil:, but could you squash the commits you have on this branch into a single commit so we can merge this?

Lussebullen commented:
@Benehiko I squashed them now, I just added my co-authors in the commit message, not sure how it works with the ctn, but its important to me that they get the credit they are due :)
